### ISSUE TITLE
Connection to remote es servers fails

### ISSUE THREAD

**Issue by tabardar** [ropensci/elastic#52]
connections to remote elasticsearch servers always fails
the following should connect but always attempts a local server

```r
connect("validipaddressonanothermachine")
connection()
# successfully returns details of the other machine running elasticsearch
ping()
# always fails with attempt to 127.0.0.1 (where there is no ES instance running)
connection()
# now shows 127.0.0.1
# the R options are somehow being overwritten
# this happens on Linux (ubuntu) and Windows (w7 64 bit) machines.
```

**Comment by sckott** [comment]
@tabardar thanks for the report. What version of `elastic` are you using?

**Comment by sckott** [comment]
@tabardar I just tried with the most recent version I have on master branch, connecting to a local version of ES running in a docker container, and it works fine. You can put the port in the url or separate them

```r
connect("<base url>", es_port = "9200")
```

or

```r
connect("<base url>:9200")
```

**Comment by tabardar** [comment]
Hello @sckott

I'm using ...
R version 3.1.3 (2015-03-09)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.2 LTS
locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] elastic_0.3.0

I can connect to a local elasticsearch with no problem and also make a connection to a remote server. The problem comes when I attempt to do anything else with the remote connection.
> connect("d630")
uri:       d630 
port:      9200 
username:  NULL 
password:  NULL 
api key:   NULL 
elasticsearch details:   
      status:                  200 
      name:                    d630 
      Elasticsearch version:   1.4.4 
      ES version timestamp:    2015-02-19T13:05:36Z 
      lucene version:          4.10.3> 

> connection()
uri:       d630 
port:      9200 
username:  NULL 
password:  NULL 
api key:   NULL 
elasticsearch details:   
      status:                  200 
      name:                    d630 
      Elasticsearch version:   1.4.4 
      ES version timestamp:    2015-02-19T13:05:36Z 
      lucene version:          4.10.3> 

> ping()
Error: 
  Failed to connect to http://127.0.0.1:9200
  Remember to start Elasticsearch before connecting

> connection()
Error: 
  Failed to connect to http://127.0.0.1:9200
  Remember to start Elasticsearch before connecting

It's almost as if the connection is being retried in the ping() command with default local parameters or perhaps the connection details in the connect() attempt are not available to the ping() command.

(keep up the good work by the way - elastic is very helpful)

**Comment by tabardar** [comment]
Sorry - accidently closed it - doh!

**Comment by sckott** [comment]
Thanks for the nice words. And sorry about the trouble. 

Can you tell me if elasticsearch is available on port 9200 on your remote IP?  Or a different port? Above you have `connect("d630")` ? What is `d630`?  I assume you put a valid IP in there? Do you put a port in the url?  or set it with the parameter `es_port` in `connect()`?

**Comment by tabardar** [comment]
Hello 

The value "d630" is the network name of a remote IP address; it translates to 192.168.1.69. The machine I am on has an IP address of 192.168.1.74. If I type "ping d630" from a linux command prompt on the .74 machine, I get responses from the .69 address. If I then type "curl d630:9200" I get a response from the Elasticsearch server from the .69 address saying that all is well.

When using elastic, I don't use the 9200 port when connecting. 
In other words when on the .74 machine 
connect(es_base="192.168.1.69", es_port="9200") makes a connection OK
then 
ping() fails and reports the local 127.0.0.1 address.

By the way, if I do this 
connect("192.168.1.69:9200")
I get an error "Failed to connect to 192.168.1.69:9200:9200"

**Comment by sckott** [comment]
Hmm, both of those IP addresses are down when I try them. Are they just temporarily down right now?

**Comment by tabardar** [comment]
They are private in my network - you would have to use two of your own machines or find a publicly available Elasticsearch server. In my setup, the .74 machine was a Linux instance running using Oracle VirtualBox inside my Windows 7 machine.

**Comment by sckott** [comment]
@tabardar oh, right, was thinking it might be a public IP I could test. 

could you try installing the latest `devtools::install_github("ropensci/elastic")` and retry, and see if you ge the same thing

**Comment by tabardar** [comment]
@sckott - I tried devtools - after some experimenting I found that connect("192.168.1.69") hangs whereas connect(es_base="http://192.168.1.69",es_port="9200) works and successfully allows the subsequent ping() to work - yay. I also found that options("es_base"="http://192.168.1.69") works with the devtools version.

**Comment by sckott** [comment]
@tabardar great!

right, you can specify `es_base` and `es_port` either through the `connect()` function, or just setting them with `options()` as `connect()` just writes to options for each R session. 

Those are stored just for the session though. If you very often use that base url, you could set it in your `.Rprofile` file

**Comment by tabardar** [comment]
@sckott  - when will the next version (I assume 0.4.0) be officially released so I can take advantage of this?

**Comment by sckott** [comment]
@tabardar planned for a few weeks from now https://github.com/ropensci/elastic/milestones - should be about then. But you can use the github version right?

**Comment by tabardar** [comment]
@sckott - yes of course - the github version will be fine - if I find any other "features" ;) I'll let you know

**Comment by KrishnaPG** [comment]
@sckott Thanks for sharing your good work - it is very helpful. 

I too faced the same situation as @tabardar and thought of sharing it here for benefit of others.

Looks like if you use `connect("192.168.56.106")` without `http` it has some problems. For example, I was able to put data into the ES using docs_bulk() add see the data in there, but none of other commands (such as docs_get, Search...) were working - they complain as below:
````
Could not resolve host: :; Host not found
````
On the other hand if we use `connect("http://192.168.56.106")` with `http://` then the commands start working great.

Perhaps emphasizing this (requirement to use `http://` prefix for the ip-address) as special note on docs may help others. 
(alternatively adding `http://` inside the connect code automatically, if user does not specify it,  is also a good option)

**Comment by sckott** [comment]
thanks @KrishnaPG  I guess you were the one that commented on the blog post too.  I just pushed a change so that `http` is added if it is missing in the base url passed in. 

I also added back the default port to be 9200 so set that to NULL if you don't need a port

**Comment by KrishnaPG** [comment]
Yes @sckott . Thanks for the changes, they are helpful andwill make it easy for new comers to adapt and do quick test.

**Comment by sckott** [comment]
:+1:
